### ISSUE TITLE
Add support for shorthand nested tags (e.g., <@foo>)

### ISSUE THREAD

**Issue by mlrawlings** [marko-js/marko#436]
The point of the layout taglib is to be able to include another template (`<layout-use>`) and insert content (`<layout-put>`) at various insertion points (`<layout-placeholder>`).  I'd like to make this more generic and concise.

@patrick-steele-idem has already done work on making `<include>` work to replace `w-body` (#418) and I think we can take this one step further to replace the need for layout-specific tags.

**Old:**
`./layouts/main/template.marko`:
```html
<!doctype>
<html>
<body>
    <layout-placeholder name="body"/>
</body>
</html>
```
`page.marko`:
```html
<layout-use('./layouts/main/template.marko')>
    <layout-put into="body">
        Hello World
    </layout-put>
</layout-use>
```

**New:**
`./layouts/main/template.marko`:
```html
<!doctype>
<html>
<body>
    <include(data.body)/>
</body>
</html>
```
`page.marko`:
```html
<include('./layouts/main/template.marko')>
    <@body>
        Hello World
    </body>
</include>
```

### `@` tags

As seen above, I'm also proposing the ability to prefix a tag with the `@` symbol to indicate that it should be passed as an attribute. I chose `@` because we already use it to indicate attributes in `marko.json` and it looks nice.

These are like [nested tags](http://markojs.com/docs/marko/custom-taglibs/#nested-tags) but more concise and don't require defining attributes in `marko-tag.json`.

Only putting the `@` at the front will require a change to the parser to support, but I think it would be worthwhile:
```html
<@body>
    Hello World
</body>
```
vs
```html
<@body>
    Hello World
</@body>
```

These `@` tags would work for *any* tag, so we could use layout concepts easily for custom tags as well:

**New:**
`./components/layout/index.marko`:
```html
<!doctype>
<html>
<body>
    <include(data.body)/>
</body>
</html>
```
`page.marko`:
```html
<layout>
    <@body>
        Hello World
    </body>
</layout>
```

### Passing values to `@` tags?

Perhaps we could also allow passing an argument to the `@` tag, indicating the name of a data variable:

```html
<my-component>
    <@body(foo)>
        Hello ${foo}!
    </body>
</my-component>
```

would compile to:
```
my-component({
    body: function(foo, out) {
        out.w('Hello '+escapeXML(foo)+'!');
    }
});
```
and could be called within `my-component` like this:
```html
<include(data.body, 'World')/>
```

**Comment by mlrawlings** [comment]
Thoughts?
/cc @philidem @austinkelleher @charlieduong94 @Hesulan

**Comment by Hesulan** [comment]
Interesting, I like it! I'm still a little fuzzy on a few of the details, though I'm also reading this on my phone and half asleep... It's been a long day.

I especially like the @ syntax. "<@body></@body>" would be fine with me if getting the parser to accept "<@body></body>" is a lot of work.

It looks like the <include> tag combined with @ tags should be able to do everything the layout taglib does and more, if so then I'd have no problem with it being deprecated in favor of this once it's fully implemented.

**Comment by Hesulan** [comment]
Bah, I think my phone messed up the formatting again... Oh well.

**Comment by mlrawlings** [comment]
The other question this raises is should we deprecate [nested tags](http://markojs.com/docs/marko/custom-taglibs/#nested-tags) in favor of `@` tags?

**Old:**
```html
<await(foo from data.fooPromise)>
    <await:error>
        Something bad happened
    </await:error>
    ${foo}
</await>
```

**New:**
```html
<await(foo from data.fooPromise)>
    <@error>
        Something bad happened
    </error>
    ${foo}
</await>
```

**Comment by Hesulan** [comment]
Would there ever be a situation where the target parent of the nested tag could be ambiguous? If not, then I can't see any reason not to deprecate the old syntax.

**Comment by Hesulan** [comment]
In other words, are there any situations where "@error" could mean either "await:error" or "something-else:error"? I can't think of any, but it's worth asking.

**Comment by mlrawlings** [comment]
I just realized that `<await:error>` isn't a thing.  It's `<await-error>` which gets transformed into an attribute at transform time... haha 😁 .  But I'm thinking we could still use `<@error>` in this case.

@Hesulan My thinking on the `<@tag>` is that they would need to be the immediate children of the target parent, so there wouldn't be any ambiguity.

**Comment by mlrawlings** [comment]
I suppose another question is how to handle multiple instances of a tag.  Currently, with nested tags, you define in the `marko-tag.json` whether a nested tag can have multiple instances.  If it can, it's always passed as an array.  

Without defining that upfront I think we're left with passing it as a single value if there's only one `<@tag>` and as an array if there are multiple.  React does this with its children prop and I don't really like it.  They ended up providing a collection of [helpers](https://facebook.github.io/react/docs/react-api.html#react.children) because it's hard to work with.

You *could* do something like the following to ensure a value is an array before looping over it:
```html
<var tabs=[].concat(data.tab)/>
<ul>
    <li for(tab in tabs)>${tab.name}</li>
</ul>
<div class="tab-content" for(tab in tabs)>
    <include(tab)/>
</div>
```

Or maybe you **have** to define an entry for nested tags if you want multiple: `"@tabs <tab>[]"`.

**Comment by Hesulan** [comment]
Hmm, that's a good question... what about a slightly different syntax for one of the two conditions? Nested tags that can be defined multiple times could use `<@tabs>`, while single-instance tags could use something like `<:tab>`.

**Comment by patrick-steele-idem** [comment]
I like this suggestion and here are some thoughts:

-----

I think it should be `<@foo>...</@foo>` (with the `@` repeated in the ending tag)

-----

Attributes should be allowed on `<@>` tags:

```xml
<@placeholder tag="div">Loading...</@placeholder>
```

-----

`<@bar>` should be shorthand for `<foo:bar>`. The only challenge is that the `<@bar>` would just need to be matched with the first ancestor custom tag:

The following are equivalent:

__Old:__

```xml
<app-tabs>
    <app-tabs:tab title="Home">Content for Home</app-tabs:tab>
    <app-tabs:tab title="Profile">Content for Profile</app-tabs:tab>
    <app-tabs:tab title="Messages">Content for Messages</app-tabs:tab>
</app-tabs>
```

__New:__

```xml
<app-tabs>
    <@tab title="Home">Content for Home</@tab>
    <@tab title="Profile">Content for Profile</@tab>
    <@tab title="Messages">Content for Messages</@tab>
</app-tabs>
```

-----

Nested tags currently compile down to an object with each attribute resulting in a property and a `renderBody(out)` function if the nested tag has body content. The nested tags get discovered at render time by rendering the body and the nested tags register themselves with the parent tag when they get rendered but they otherwise produce no output at that time. This allows nested tags to be repeated, have conditional logic, etc.:

```
<app-tabs>
    <for(person in family)>
        <@tab title=person.name>
            Hello ${person.name}! You are ${person.age} year(s) old.
        </@tab>
    </for>
</app-tabs>
```

However, we can optimize this by handling nested tags that are direct children of a parent tag:

```xml
<app-hello name="Frank">
  <@message>Have a good day!</@message>
</app-hello>
```

__Old compiled output:__

```javascript
app_hello_tag({
      name: "Frank"
  }, out, 0, function renderBody(out, app_hello_tag0) {
    app_hello_message_tag({
        renderBody: function renderBody(out) {
          out.w("Have a good day!");
        }
    }, out, app_hello_tag0);
  });
```

__New compiled output:__

```javascript
app_hello_tag({
        name: "Frank",
        message: {
            renderBody: function renderBody(out) {
                out.w("Have a good day!");
            }
        }
    }, out);
```

We could even optimize for a single `if()` condition surrounding the nested tag:

```xml
<app-hello name="Frank">
  <@message if(someCondition)>Have a good day!</@message>
</app-hello>
```

```javascript
app_hello_tag({
        name: "Frank",
        message: someCondition && {
            renderBody: function renderBody(out) {
                out.w("Have a good day!");
            }
        }
    }, out);
```

-----

__tldr; Introduce`<@tab>` as shorthand for `<tabs:tab>`__
